Mas i1775 ttaaesafety

martinsumner opened the pull request:
Bug fixes and improvements to TTAAE full-sync

martinsumner committed:
Add backoffs on query timeouts

Re-calibrate defaults

martinsumner committed:
Switch to using fetch_clocks_range

This also appears to be much quicker within `riak_test`.  Why, is not clear.  Perhaps related to queueing at aae_runner as perfomance log within aae_runner is fast even when fetch_clocks_nval returns a slow response.

This change is leading towards the possibility of adding modified_range to fetch_clocks_nval, and so allowing for hourcheck and daycheck as well as allcheck in ttaaefs schedules.

martinsumner committed:
Fix issue with format switches onmodifed date ranges

remote and local clients have alternate expectation on whether `data` included as first elemen tof tuple.

martinsumner committed:
Switch kv_index_tictactree branch

To allow for option to add modified range on "full" exchanges

martinsumner committed:
Add support for DayCheck/HourCheck on full fullsync

martinsumner committed:
Add API changes for fetch_clock_nval

To allow for modified ranges on nval_checks

martinsumner committed:
Correction to default nval filter

Is expecting a large tree (as cached trees are always large)

martinsumner commented:
https://github.com/basho/riak_test/compare/develop-3.0...mas-i1775-ttaaesafety

martinsumner committed:
Schema and log refinements

By default have extra capacity for aae_folds used in replication.

Pretty print buckets and keys in repair logs where possible.

martinsumner committed:
Temp commit

Why do deletes not get reaped - but only on one site

martinsumner committed:
Add configurable pause

There exists a race condition here, and wan to try and take the minimal step to mitigate this condition.  There is not hurry at this point, as the client has already been informed of the success/failure of the delete.

Assumption is that a pause approximately equal to the intra-cluster network delay will be sufficient.

Note that the race can be cleared up even if it happens (e.g. by repl or by manual reap).  Aim is to contorl the volume at which it happens (to make it easier for repl to clear up without intervention of manual repl).

martinsumner committed:
Revert "Temp commit"

This reverts commit 33a963bd55ae97810bae07a1753766238bee6fd9.

martinsumner committed:
Improve reporting of repairs

Make this log more operator friendly - especially if an operator may need to run a range_repl to recover.

martinsumner committed:
Update rebar.config

martinsumner committed:
Fix repair logging

martinsumner committed:
Use environment variable to boost/quiet work items

When a work item doesn't find much repair work, then quieten the max results so that it does less work.  This should allow for more frequent scheduling (e.g. without having lots of work initiated from a cluster which is behind).

martinsumner committed:
Log and quietener improvements

martinsumner committed:
split queues for all_sync and modified_range queries

Also increase timeout for exchanges to 1 hour (half the 2 hour crash timeout)

martinsumner committed:
Decode clock before compare

Otherwise compare will all be done in repair - and take a long time

martinsumner committed:
Decode clocks in mod_range queries

martinsumner committed:
Remove auto-quitener

This may well be the right thing to do (quiten the more complicated queries) - but the resultant behaviour isn't intuitive.

martinsumner committed:
Add support for range_check

Replaces the boost concept.  Now range_checks can be added to the schedule, and when a delta is discovered the range of the delta will be set to be used in the next range_check query.

If there is no range set, then the range_check is a null op.  The range is always unset before each range_check, and will only be reset if sufficient results are found.

The expectation is that fewer hour/day/all checks should need to be configured as the bulk of the work will be managed by range_check.  As the range_check is handling an identified problem - it will be run with boosted max_results.

martinsumner committed:
Add scheduling support for range_sync

Also expose the set/get/clear functions for the range - make it easier to control this from a remote_console and/or tests.

Need to account for the fact that rnage queries may now be requested on NVal tests - and so NVal may be an integer not range on fetch_clocks_range.

martinsumner committed:
Resolve name clash

martinsumner committed:
Correct Mesgaseconds

Also

- splitting full-sync folds onto different queues was confusing, and perhaps unnecessary with range_check.  Revert this back, and expand the default size of the af3 queue.

- use the kv_index_tictcatree feature to log purpose of exchanges, to more easily differrentiate between inter and intra cluster AAE in logs

martinsumner committed:
Typo

martinsumner committed:
Make use of range_check

When AAE is working (i.e. clusters in sync) still use range_check, just sync since the last success.  Will find ranges quicker after short outages.

If there has not been success, and there has not been a range set (i.e. this site is behind), then don't do range_check.  The range_check will have to wait for another check to either succeed, or set a range so that range_checks can continue.

martinsumner committed:
Correct success reply names

martinsumner committed:
Should assume a success on startup

Otherwise very low to start AAE exchanges

martinsumner committed:
Improve handling of crashed nodes

Following improvements:

Catch local sender errors (e.g. insufficient_vnodes_available)

Put rebuilds (following crashes) onto a separate configurable queue

Allow for a riak_client function to take recovering nodes out of participate_in_coverage

martinsumner committed:
Add log on repairing/failed exchange

martinsumner committed:
Use _check not _sync consistently in schema and code

Was getting confusing

martinsumner committed:
Use a specific bucket range check if hits target

This has been shown to be potentially much quicker if there are multiple buckets that need repairing

martinsumner committed:
Use more cautious defaults

This as before backoff kicks in, there cna be a short term impact of lots of AAE on larger nodes (in terms of key count) - and we should assume large key counts are the default.

martinsumner committed:
Update riak_kv.schema

martinsumner committed:
Check the bucket with the biggest key count

martinsumner committed:
Update rebar.config

martinsumner committed:
Add compaction run control into leveled backend

martinsumner committed:
Make pause in exchange configurable

Is this related to occasional false deltas under stress?

martinsumner committed:
ttaaefs_manager - not too clever

Trying to pick a single bucket more often was useful in some tests, however it could lead to strange results as the "top" bucket drained of results (especially if mroe than one range_check picked up the last gasp.  This led to more lumpy resolution.

The repairing behaviour has a big impact on performance, and cna be turned on if required.  Leave repair to intra-cluster AAE.  don't duplicate this effort.

martinsumner committed:
Update NextGenREPL-GettingStarted.md

ThomasArts commented on priv/riak_kv.schema:
```diff
@@ -90,14 +90,14 @@
-%% maybe prompt the next exchange. Default is 4 minutes - check all partitions
-%% when n=3 every 30 minutes.
+%% maybe prompt the next exchange. Default is 8 minutes - check all partitions
+%% when n=3 once every hour (in each direction).
```
probably add the formula on how to get from 8 and 3 to once every hour. That is not straigthforward

martinsumner replied:
Added - it is 1 + (n - 1) * n ticks per cycle.  there are n - 1 exchanges for each partition, and n partitions ... plus 1 tick to prompt the refill of the cycle

ThomasArts commented on priv/riak_kv.schema:
```diff
@@ -852,6 +852,33 @@
+%% If using range_check to speed-up repairs, this can be reduced as the
+%% range_check maxresults will be boosted by the ttaaefs_rangeboost  When using
+%% range_check a value of 64 is recommended, reducing to 32 or 16 if the cluster
+%% has very a large volume of keys and limited capacity.  Only reduce below
```
```suggestion
%% has a very large number of keys and limited capacity.  Only reduce below
```

ThomasArts commented on priv/riak_kv.schema:
```diff
@@ -852,6 +852,33 @@
+%% @doc For tictac full-sync what is the maximum number of AAE segments to be
+%% compared per exchange.  When running a range_check query this will be the
+%% ttaaefs_max results * ttaaefs_rangeboost.
+%% When using range_check, a small maxresults can be used, ineffecting making
```
Don't understand `infecting making`

martinsumner replied:
Corrected

ThomasArts commented on src/riak_client.erl:
```diff
@@ -878,7 +879,35 @@ ttaaefs_fullsync(WorkItem, SecsTimeout, Now) ->
 
-
+-spec remove_node_from_coverage() -> ok.
+remove_node_from_coverage() ->
```
If you give an argument to this function ParticipateInCoverage, then the second one can call the first with the default value. Less code, more consistency

ThomasArts commented on src/riak_kv_delete.erl:
```diff
@@ -94,6 +97,7 @@ delete(ReqId,Bucket,Key,Options,Timeout,Client,ClientId,VClock) ->
                     ?DTRACE(?C_DELETE_INIT2, [1], [<<"reap">>]),
                     {ok, C2} = riak:local_client(),
                     AsyncTimeout = 60*1000,     % Avoid client-specified value
+                    timer:sleep(TombPause),
```
Explain, why is 10 milliseconds a good time to resolve the issues you have?

martinsumner replied:
The delay is harmless, in the sense the client has already had a response (although it will slow down deletes queued by riak_kv_eraser).

I suspect 1ms will make it go away, there'a network hop involved in the race condition.  However, other environments might be different.  

It is configurable if it proves to be a problem, so will leave at 10ms until more data comes back from deployments.

ThomasArts commented on src/riak_kv_vnode.erl:
```diff
@@ -1686,15 +1702,44 @@ handle_aaefold({merge_branch_nval, Nval, BranchIDs},
+            %% used for scheduling the work not the aae_runner.  Also, the 
+            %% fetch clock query will not attempt to rebuild segments of the
+            %% tree cache - this is left to internal anti-entropy
+            handle_aaefold({fetch_clocks_range,
```
kind of too many arguments to be able to really get what's going on here.

ThomasArts replied:
you could equally well have called with fetch_clocks_nvals, since that directly returns in this new call.

martinsumner replied:
This way just helps with concurrently managing the update to 2.9 and 3.0, as 2.9 won't initially have the new fetch_clocks_nval option

ThomasArts commented on src/riak_kv_ttaaefs_manager.erl:
```diff
@@ -345,53 +385,146 @@ handle_cast({all_sync, ReqID, From, _Now}, State) ->
             % The bigger the time range, the bigger the tree.  Bigger trees
             % are less efficient when there is little change, but can more
-            % accurately reflect bigger changes (with less false positives).
+            % accurately reflect bigger changes (with less fasle positives).
```
```suggestion
            % accurately reflect bigger changes (with less false positives).
```

ThomasArts commented on src/riak_kv_vnode.erl:
```diff
@@ -1111,14 +1111,23 @@ handle_command({rebuild_complete, trees, _ST}, _Sender, State) ->
+        UnwelcomeState ->
+            lager:info("Tictac AAE exchange for partition=~w pending_state=~w",
+                        [State#state.idx, UnwelcomeState])
+    end,
```
Not sure you like list comprehensions, but for such a side-effect: do this if Condition, they are really nice.
The above code would become a one-liner:
```
     [ lager:info("Tictac AAE exchange for partition=~w pending_state=~w",
       [State#state.idx, EndState]) || EndState /= root_compare andalso EndState /= branch_compare].
```
Instead of `andalso` you can use `,` or `and`.
But you may also absolutely dislike it :-).

martinsumner replied:
I'm more concerned that I've leaked some knowledge of an atom in kv_index_tictactree into riak_kv.  I think really I should have a `aae_exchange:positive_outcomes/0` exported, so that riak_kv doesn't need to know about what these mean, and is decoupled from new positive outcomes being added.

I will look at this.

martinsumner replied:
Having fiddled, I'm only making matters worse -swapping one bit of special knowledge for another.  will leave as-is

ThomasArts commented on src/riak_kv_ttaaefs_manager.erl:
```diff
@@ -52,11 +56,17 @@
     % prompting another
--define(CRASH_TIMEOUT, 3600 * 1000).
+-define(CRASH_TIMEOUT, 7200 * 1000).
     % Assume that an exchange has crashed if not response received in this
```
```suggestion
    % Assume that an exchange has crashed if no response received in this
```

ThomasArts commented on src/riak_kv_ttaaefs_manager.erl:
```diff
@@ -345,53 +385,146 @@ handle_cast({all_sync, ReqID, From, _Now}, State) ->
+                        NowSecs = MegaSecs * ?MEGA  + Secs,
+                        {all,
+                            all,
+                            PrevMega * ?MEGA + PrevSecs - 60,
```
This confuses me. In the hourly and daily, the Now argument is reduced by nr of seconds in day/hour respectively.
But here we subtract seconds from the previous successful sync. Should we not have something like, between max(PreviousSuccess , Now - 60), Now ?

martinsumner replied:
The 60 is a bit of a fuzz factor, that part-fixes a problem that the last success was prompted potentially a bit before this timestamp.  I'll either get the correct timestamp passed through in some way, or make it replace 60 with ?FUZZ with a comment on the definition of FUZZ to explain

martinsumner replied:
Actually, no, I was wrong.  The last success time is the time the last successful query was prompted.  So I think the fuzz factor is unnecessary

ThomasArts commented on src/riak_kv_ttaaefs_manager.erl:
```diff
@@ -345,53 +385,146 @@ handle_cast({all_sync, ReqID, From, _Now}, State) ->
+                    {State0, Timeout} =
+                        sync_clusters(From, ReqID, range, range, Filter,
+                                        NextBucketList, partial, State,
+                                        hour_check),
```
hour_check or range_check here?

martinsumner replied:
should be range_check

ThomasArts commented on src/riak_kv_ttaaefs_manager.erl:
```diff
@@ -617,8 +835,25 @@ remote_sender({fetch_clocks, SegmentIDs}, Client, Mod, ReturnFun, NVal) ->
             {ok, {keysclocks, KeysClocks}} ->
-                ReturnFun(lists:map(fun({{B, K}, VC}) -> {B, K, VC} end,
-                            KeysClocks));
+                MapFun =
```
you use this MapFun at a lot of places. Would it not be better to write a function
`decode_keyclocks` that takes a list and does exactly this.... less code duplication

martinsumner replied:
Updated

ThomasArts commented on src/riak_kv_ttaaefs_manager.erl:
```diff
@@ -725,13 +962,45 @@ generate_repairfun(ExchangeID, QueueName) ->
+                                    [B, LowDT, HighDT, KC, WorkType]),
+                        set_range(B, all, LowDT, HighDT);
+                    _ ->
+                        MapFun = fun({_B, _KC, LD, HD}) -> {LD, HD} end,
```
if you only want a LoDT and HiDT, then this code can be replaced by a fold.
```
[{_,_, LD1, HD1}|_] ->
   {LoDT, HiDT} = 
      lists:foldl(fun({_B, _KC, LD, HD}, {Min, Max}) ->
                               {min(LD, Min),
                               {max(HD,Max)}
                 end, {LD1,HD1}, PBDL),

ThomasArts replied:
I this way yo don't need the case either, because if you have one element it's just a special case

martinsumner replied:
I'm OK with how it reads at the moment.  I'm minded to leave as-is for now.

ThomasArts commented on src/riak_kv_ttaaefs_manager.erl:
```diff
@@ -807,41 +1110,57 @@ take_next_workitem([NextAlloc|T], Wants,
+    lists:sort(
+        choose_schedule(Slices,
+                        Allocations,
+                        {NoSync, AllSync, DaySync, HourSync, RangeSync})).
```
naming may be a bit weird here, since you renamed other to NoCheck, AllCheck, etc.

martinsumner replied:
Changed to *Check

martinsumner committed:
Tidy-up post review

martinsumner committed:
BE AWARE - potential breaking change

Tombstones do not currently have a last modified date.  This has not caused any issues, as last modified dates are only checked:

https://github.com/basho/riak_kv/blame/d17495741ed001dcb54fa9e75e0dff026baaefde/src/riak_kv_vnode.erl#L3048-L3049

This would have previously blown up before the relative safe get_last_modified function was added, as opposed to doing a dict:fetch for the tag (which would have triggered an exception).

This is necessary to have time-based rpelication work correctly.  Currently there can be issues where recent tombstone discrepancies cna only be picked up by all_check (even if they are recent).  This is confusing.

Should we just live with the confusion rather than risk changing this?  But it feels like tomstones not having a LMD is just a mistake i.e. it being useful is just a scenario which was missed, and there are no issues.

Note, as a side effect, tombstones also get a vtag.

Perhaps some client somwhere will find this weird/unexpected?

martinsumner committed:
Typo

martinsumner reviewed (commented):
There is at least a switch to disable this {riak_kv, tombstone_timestamp} can be set to false - so anyone having an issue can simply flick that switch.

martinsumner committed:
Switch to improve list_buckets

martinsumner committed:
Changes following review

martinsumner committed:
Additional review changes

Bob-The-Marauder commented:
Not sure if this comment is entirely relevant but including it just in case.

As part of system tuning, it's recommended to set the disks to `noatime` mounts https://www.tiot.jp/riak-docs/riak/kv/2.2.6/using/performance/#mounts

I know the question in https://postriak.slack.com/archives/C6R0LPH4N/p1607107910139900 uses "last_modified_date" and that `noatime` is only relevent for "last accessed" but thought I should mention it in case there is any potential for people to  tune Riak and, in doing so, inadvertently break functions.

martinsumner commented:
@Bob-The-Marauder - last modified time is the time on an individual object (which comes from the OS system time), and not related to times on files etc.  So this won't be an issue.

martinsumner committed:
Export validity checks for aae folds

Otherwise invalid queries get no repsonse to riak_client.

martinsumner committed:
Shortcuts to riak_client functions

martinsumner committed:
Missing export

martinsumner committed:
Switch to develop-2.0 branches

martinsumner committed:
Update NextGenREPL-GettingStarted.md

martinsumner committed:
Update rebar.config
